sstables, replica: add generation generator

tchaikov opened the pull request:
this is the first step to the uuid-based generation identifier. the goal is to encapsulate the generation related logic in generator, so its consumers do not have to understand the difference between the int64_t based generation and UUID v1 based generation.

this commit should not change the behavior of existing scylla. it just allows us to derive from `generation_generator` so we can have another generator which generates UUID based generation identifier.

tchaikov commented:
@bhalevy hi Benny, this is the generator i mentioned over the standup. i have a branch locally which is based on your two `generation_type` related PRs. but wanted to get some early inputs on this change first.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4724/

avikivity commented on sstables/generation_type.hh:
```diff
@@ -36,6 +38,22 @@ constexpr int64_t generation_value(generation_type generation) {
+};
+
+class int_generation_generator final : public generation_generator {
+    std::atomic<int64_t> _last_generation;
```
Why atomic?

We may need to keep the last int generation even if it's not in use, if we allow configuring the generation type back to int.

tchaikov replied:
do we really support changing the generation id back to int? thought there is no u turn here.

tchaikov replied:
> > Why atomic?
> 
> good question. i asked this to myself before copying it from the existing code. just wanted to preserve the existing implementation as much as possible, and focus on the generalizing it with a generator class, and keep the cleanups in a separate commit(s).
> 

explained in the comment in code. in short, this is due to how reshard is implemented -- we generate generation identifier from another shard on behalf of the owner shard of a given sstable.

denesb commented on replica/distributed_loader.cc:
```diff
@@ -401,9 +401,38 @@ distributed_loader::make_sstables_available(sstables::sstable_directory& dir, sh
+                                 gc_clock::now(),
+                                 &error_handler_gen_for_upload_dir);
+    }
+};
```
Can't we fold this into the `sstable_manager` instead? It is already mandatory to create sstables.

tchaikov replied:
good point. will implement this and fix the test failures in the next revision !

tchaikov replied:
`sstable_manager` does not track the latest generation so far. and there are multiple different use cases where the generation is passed to `make_sstable()`. i am not sure how to consolidate them without sacrificing the flexibility. could you be more specific?

denesb replied:
I just dislike creating a brand new sharded service just for generating generations. Its one more service we have to propagate down to every place which wants to create sstables. As the sstables-manager is already mandatory on all these sites, I though it could also take on the role of generating the generations. I don't have anything more specific in mind than that, sorry.

tchaikov replied:
i see. and i concur with you. the sharded service is overkill and does not fit in this use case. as it is not symmetric. and the states are shared across multiple shards. the latest revision of this PR just uses a degraded form of generation identifier generator.

raphaelsc replied:
Very good idea on moving it to sstable manager. replica::table talks to sstable manager for everything, including picking a new sstable with a new generation.

today, we have:

sstables::shared_sstable table::make_sstable(sstring dir) {
    auto& sstm = get_sstables_manager();
    return sstm.make_sstable(_schema, dir, calculate_generation_for_new_table(), sstm.get_highest_supported_format(), sstables::sstable::format_types::big);
}

but we can change it to become:

sstables::shared_sstable table::make_sstable(sstring dir) {
    auto& sstm = get_sstables_manager();
    return sstm.make_new_sstable(_schema, dir);
}

tchaikov replied:
i think the change you are proposing implies that we can move `calculate_generation_for_new_table()` into `sstables_manager`, but we only hsave two sstables_managers in each scylla server, one for user tables, another for system tables. so, this in turn has an assumption that the highest generation id can be shared across different tables, and it should be stored in `sstables_manager`. but currently, each table keeps track of its own generation id, and it uses it to tell if it is ready for write. so, i am afraid that we cannot store the highest generation id in `sstables_manager`, neither can we use `sstables_manager` for creating new generation id.

one of the goals of generation generator design is to avoid segmenting the implementations of how we generation id generation, and to consolidate them to a single place. so, i think we might need to keep 
```c++
sstables::shared_sstable table::make_sstable(sstring dir) {
    auto& sstm = get_sstables_manager();
    return sstm.make_sstable(_schema, dir, 
        calculate_generation_for_new_table(),  
        sstm.get_highest_supported_format(),
        sstables::sstable::format_types::big);
}
```

as it is.

tchaikov commented:
cc @raphaelsc

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4834/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4838/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4857/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4891/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4892/

tchaikov commented:
```
13:55:57  time="2023-03-12T07:55:56+02:00" level=warning msg="Error validating CNI config file /jenkins/.config/cni/net.d/87-podman.conflist: [failed to find plugin \"bridge\" in path [/usr/local/libexec/cni /usr/libexec/cni /usr/local/lib/cni /usr/lib/cni /opt/cni/bin] failed to find plugin \"portmap\" in path [/usr/local/libexec/cni /usr/libexec/cni /usr/local/lib/cni /usr/lib/cni /opt/cni/bin] failed to find plugin \"firewall\" in path [/usr/local/libexec/cni /usr/libexec/cni /usr/local/lib/cni /usr/lib/cni /opt/cni/bin] failed to find plugin \"tuning\" in path [/usr/local/libexec/cni /usr/libexec/cni /usr/local/lib/cni /usr/lib/cni /opt/cni/bin]]"
```
env issue.

xemul commented on replica/distributed_loader.cc:
```diff
@@ -406,6 +406,26 @@ sstables::shared_sstable make_sstable(replica::table& table, fs::path dir, int64
 }
 
+class generation_generator {
+    // We still want to do our best to keep the generation numbers shard-friendly.
```
Please, note, that the fully-qualified name would then look like `sstables::sstable_generation_generator`

tchaikov replied:
why? i think we don't need to add sstables in the symbols in "sstables" namespace, just like, in general, we don't add "seastar" prefix in the symbols in "seastar" namespace.

xemul replied:
> i think we don't need to add sstables in the symbols in "sstables" namespace

In fact we do. There are

* `sstables::sstable`
* `sstables::sstables_manager`
* `sstables::sstable_directory`
* `sstables::sstable_set`
* `sstables::sstable_version_types`
* `sstables::sstable_writer_config`

and more

tchaikov replied:
fair enough. will add it back.

raphaelsc replied:
the idea of the prefix is to make it explicitly for someone that it's a type inherited from sstable. for example, in sstable namespace, there might be helpers that are not necessarily part of what composes a sstable.

raphaelsc replied:
I understand it's somewhat redundant, as one can think everything in sstables namespace should be related to sstable to some extent, but having the prefix also adds some clarification when we don't explictly specify the sstables namespace when declaring types from within the sstables namespace itself.

tchaikov replied:
@raphaelsc hi Raph, thanks for the explanations. with/without the prefix do have their pros and cons. but consistency in naming is more important than a shorter name. i will revisit this topic later.

tchaikov commented:
changelog

- s/generation_generator/sstable_generation_generator/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4902/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4904/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4906/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4921/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4931/

denesb commented on replica/distributed_loader.cc:
```diff
@@ -412,6 +413,22 @@ sstables::shared_sstable make_sstable(replica::table& table, fs::path dir, int64
+    // so we have to use atomic<> here.
+    std::atomic<int64_t> _last_generation;
+public:
+    sstable_generation_generator(int64_t last_generation)
```
explicit

denesb commented on replica/distributed_loader.cc:
```diff
@@ -452,12 +458,12 @@ distributed_loader::process_upload_dir(distributed<replica::database>& db, distr
         process_sstable_dir(directory, flags).get();
 
-        std::vector<std::unique_ptr<sstable_generation_generator>> gen;
+        std::vector<std::unique_ptr<sstable_generation_generator<true>>> gen;
```
Please use a tagged type, looking at the code here it is impossible to guess what this `true` means.
You can use something like `use is_shared = bool_class<struct is_shared_tag>;`, then, you will have `is_shared::yes` here, which is much more readable.

tchaikov replied:
i am afraid that we cannot use an instance of `bool_class` as the template parameter. as it is not a "structural type", see https://en.cppreference.com/w/cpp/language/template_parameters -- it has non-static member variable. so i just added a 

```c++
struct is_shared {
     using yes = std::true_type;
     using no = std::false_type;
};
```

just for better readability instead of enforcing using yet another type by defining a new type using `std::integral_constant` or having a template class inheriting from `bool_constant<B>`. both approaches look like overkill.

denesb commented on db/config.cc:
```diff
@@ -826,6 +826,9 @@ db::config::config(std::shared_ptr<db::extensions> exts)
     , enable_sstables_md_format(this, "enable_sstables_md_format", value_status::Unused, true, "Enable SSTables 'md' format to be used as the default file format.  Deprecated, please use \"sstable_format\" instead.")
     , sstable_format(this, "sstable_format", value_status::Used, "me", "Default sstable file format", {"md", "me"})
+    , enable_uuid_sstable_identifiers(this,
+            "uuid_sstable_identifiers_enabled", value_status::Used, false, "If set to true, each newly created sstable will have a UUID "
```
Name should match variable name.

tchaikov replied:
okay. will incorporate this suggestion in another PR.

denesb replied:
When I see a field in the `scylla.yaml`, I come here and search for that exact name and expect to find the member variable holding that value. Not finding it like that would be terribly confusing and annoying too. I think I'm not alone doing this.

tchaikov replied:
yeah. makes sense to me. this saves a hop in searching. and developer's time is precious.

denesb reviewed (changes requested):
I think the last two commit should be removed from this series, they should be added to a PR which then actually uses the cluster feature and config item. Otherwise we risk merging a config item and cluster feature that remains unused.

The rest LGTM, minor comments only.

tchaikov commented:
changelog

- add `explicit` to the ctor of `sstable_generation_generator`
- pass `bool_constant` as the template parameter of `sstable_generation_generator` for better readability
- dropped the last two commits, as they adds the option and cluster feature bit. they are not yet wired to the generator yet.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/30/

tchaikov commented on sstables/generation_type.hh:
```diff
@@ -36,6 +37,33 @@ constexpr int64_t generation_value(generation_type generation) {
+        return sstables::generation_from_value(v + shard);
+    }
+};
+
```
> what's your plan when extending it to UUID?
> 
> IsShared template parameter will be bogus for UUID, right?
> 

yeah, the `IsShared` setting is a compile-time configuration, while UUID/integer is a runtime setting. the implementation would look like https://github.com/tchaikov/scylladb/commit/c50d7bf98e099a1f6bcdf901acbca7c6bd230572#diff-2019b4ff0737c2b01c33f32fbdb5e54fa68b98af76250fc7307e58ff2de5b613R22

> I am thinking if it it wouldnt be simpler to just stick to std::atomic, as the frequency of generator being called correlates with a new sstable creation, and that happens only once in >1 seconds. The difference for simple arithmetic is in the order of very few microseconds, so the overhead due to atomicity guarantees won't really affect the overall performance
> 

yeah, good point indeed. but i think it'd be better to be more explicit, as we are working on a project which enforces a share-nothing design. whenever one sees `atomic<>`, he/she would feel uncomfortable and starts asking "why?". and personally, i don't think it matters much. the only place we need to use a "shared" generator is when we do reshard. once the sstables find their home and are settled, we are back to a share-nothing world.

> As a benefit, transition to UUID is easier without the template parameter.

we could either have a default template parameter or define an alias for the more popular one -- the non-shared one.

bhalevy replied:
> we could either have a default template parameter or define an alias for the more popular one -- the non-shared one.

Or instantiate a `sharded<>` one where needed and use invoke_on instead of using shared std::atomic state.

denesb replied:
@bhalevy we already have a sharded generator in effect, the problem is that callers call from across shards.

denesb replied:
Why not put the generator in `generation_type.hh` in the first place?

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/48/

tchaikov commented:
rebased on to #12994 , and converting to draft.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/284/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/298/

tchaikov commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/298/

#8079

scylladb-promoter commented:
CI state **`ABORTED`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/301/

scylladb-promoter commented:
CI state **`ABORTED`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/303/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/343/

avikivity commented:
I didn't review, but does something as simple as bumping a counter or calling a random uuid generator require 15 commits? It seems excessive. Why is it so complicated?

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/346/

tchaikov commented:
> I didn't review, but does something as simple as bumping a counter or calling a random uuid generator require 15 commits? It seems excessive. Why is it so complicated?

@avikivity it is based on #12994, which includes a series of cleanup. hence this one has 15 commits.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/406/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/410/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/411/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/419/

tchaikov commented:
changelog

- rebase against master HEAD to resolve the conflicts
- dropped the commits adding cluster feature bit and the uuid generation id option
- add a commit to use `generate_n()` in the sstables test which prepares for using the generation generator

@denesb hi Botond, i just rebased the changes against master HEAD. could you take another look? i have some commits which wire the cluster feature bit to the generation generator, but before adding them on top of this change set, i just wanted to make sure that i am in the right direction and we are on the same page.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/429/

bhalevy commented on test/boost/database_test.cc:
```diff
@@ -409,14 +409,12 @@ SEASTAR_THREAD_TEST_CASE(test_distributed_loader_with_pending_delete) {
-    size_t num_gens = 9;
-    gen.reserve(num_gens);
-    for (auto i = 0; i < num_gens; i++) {
+    size_t NUM_GENS = 9;
```
nit: by convention we use all caps for constant definitions.
I'd keep `num_gens` here in lower case (maybe it should even be randomized as a follow up).
Currently, as a constant it be be defines as `constexpr`.

tchaikov replied:
okay. i don't feel strong either way. just felt that it's a constant. will

1. keep it lowercase
2. add `constexpr`

tchaikov commented on replica/distributed_loader.cc:
```diff
@@ -406,6 +407,22 @@ sstables::shared_sstable make_sstable(replica::table& table, fs::path dir, sstab
+    //
+    // operator() is called by multiple shards in parallel when performing reshard,
+    // so we have to use atomic<> here.
+    std::atomic<int64_t> _last_generation;
```
> s/int64_t/sstables::generation_type::int_t/

i understand the rationales behind this. the only thing i don't like is that it's too lengthy: 7 chars versus 32 chars. i will change this commit anyway.

> 
> Or better keep (non-atomic) `sstables::generation_type` and use `submit_to` (or `invoke_on` if we instantiate a `sharded<generation_generator>`) to manipulate it, calling `replica::table::make_new_generation` to generate new generation so we have a single source for this logic.

the idea of `sharded<generation_generator>` was considered in a previous revision of this PR. but it was considered overkill by both Avi and Botond. see https://github.com/scylladb/scylladb/pull/13073#discussion_r1130840951 (i cannot find Avi's comment on this, it was on another PR). and i agree with them that the generation generator should not a sharded service. not because it cannot be, just because the complexity of it does not deserve a new service. i plan to drop  `replica::table::make_new_generation` in a follow up PR, as it is only used by tests where a generation generator would suffice. another reason to drop this static function is, by then, we would need to access `_sstables_manager.feature_service().uuid_sstable_identifiers` when generating a new generation. one way to get around this to pass the generation generator, last known generation and `uuid_sstable_identifiers` to this static method, but this renders this method a very thin wrapper of the generator, and its whole point is to call it with other parameters.

bhalevy replied:
> > s/int64_t/sstables::generation_type::int_t/
> 
> i understand the rationales behind this. the only thing i don't like is that it's too lengthy: 7 chars versus 32 chars. i will change this commit anyway.

Hopefully it will go away, or remain in a few select places after the transition to uuid-based sstable generations.

> > Or better keep (non-atomic) `sstables::generation_type` and use `submit_to` (or `invoke_on` if we instantiate a `sharded<generation_generator>`) to manipulate it, calling `replica::table::make_new_generation` to generate new generation so we have a single source for this logic.
> 
> the idea of `sharded<generation_generator>` was considered in a previous revision of this PR. but it was considered overkill by both Avi and Botond. see [#13073 (comment)](https://github.com/scylladb/scylladb/pull/13073#discussion_r1130840951) (i cannot find Avi's comment on this, it was on another PR). and i agree with them that the generation generator should not a sharded service. not because it cannot be, just because the complexity of it does not deserve a new service.

I thought of just using the sharded instance in the distributed loader to get away from std::atomic across shards.
This is not a performance-intensive path so I'd prefer clarity of the code over performance / misleading simplicity.

> i plan to drop `replica::table::make_new_generation` in a follow up PR, as it is only used by tests where a generation generator would suffice. another reason to drop this static function is, by then, we would need to access `_sstables_manager.feature_service().uuid_sstable_identifiers` when generating a new generation. one way to get around this to pass the generation generator, last known generation and `uuid_sstable_identifiers` to this static method, but this renders this method a very thin wrapper of the generator, and its whole point is to call it with other parameters.

okay, looking forward to it :)

tchaikov replied:
> > > s/int64_t/sstables::generation_type::int_t/
> > 
> > 
> > i understand the rationales behind this. the only thing i don't like is that it's too lengthy: 7 chars versus 32 chars. i will change this commit anyway.
> 
> Hopefully it will go away, or remain in a few select places after the transition to uuid-based sstable generations.
> 

yeah, we still have some tests where we hardwire the type to integer. but yeah, i think we will fix them if they don't really need to use integer. i think the reason they use integer is just that it was convenient.

> > > Or better keep (non-atomic) `sstables::generation_type` and use `submit_to` (or `invoke_on` if we instantiate a `sharded<generation_generator>`) to manipulate it, calling `replica::table::make_new_generation` to generate new generation so we have a single source for this logic.
> > 
> > 
> > the idea of `sharded<generation_generator>` was considered in a previous revision of this PR. but it was considered overkill by both Avi and Botond. see [#13073 (comment)](https://github.com/scylladb/scylladb/pull/13073#discussion_r1130840951) (i cannot find Avi's comment on this, it was on another PR). and i agree with them that the generation generator should not a sharded service. not because it cannot be, just because the complexity of it does not deserve a new service.
> 
> I thought of just using the sharded instance in the distributed loader to get away from std::atomic across shards. This is not a performance-intensive path so I'd prefer clarity of the code over performance / misleading simplicity.
> 

i like the idea. at first glance, the `atomic<>` is always a signal of bad smell from a scylla developer's perspective. i think we can address this in a follow up change once we all agree on this. @xemul , hi Pavel, what do you think?

> > i plan to drop `replica::table::make_new_generation` in a follow up PR, as it is only used by tests where a generation generator would suffice. another reason to drop this static function is, by then, we would need to access `_sstables_manager.feature_service().uuid_sstable_identifiers` when generating a new generation. one way to get around this to pass the generation generator, last known generation and `uuid_sstable_identifiers` to this static method, but this renders this method a very thin wrapper of the generator, and its whole point is to call it with other parameters.
> 
> okay, looking forward to it :)

tchaikov commented:
changelog

- s/NUM_GENS/num_gens/ and add `constexpr` specifier
- s/int64_t/sstables::generation_type::int_t/

@denesb @bhalevy hi Botond and Benny, could you help re-review this changeset?

bhalevy commented on replica/distributed_loader.cc:
```diff
@@ -434,28 +451,24 @@ distributed_loader::process_upload_dir(distributed<replica::database>& db, distr
+        std::generate_n(std::back_inserter(shard_gen),
+                        smp::count,
+                        [highest_generation] {
+                            uint64_t base = (highest_generation.value() / smp::count + 1) * smp::count;
```
it doesn't seem like you need to `+ 1` here since `sstable_generation_generator::operator()` adds smp::count to the last generation

tchaikov replied:
i don't think so. what `sstable_generation_generator::operator()` returns is practically
```c++
_last_generation.fetch_add(smp::count, std::memory_order_relaxed) + this_shard_id()
```

and `fetch_add()` returns the original value before being incremented by `smp::count`. and before my change, the integer based generation id starts with `smp::count` , assuming the highest_generation is 0. as we always add 1 to the base before multiply it by `smp::count`.

bhalevy replied:
yup, you're right. sorry for the noise

bhalevy commented on replica/distributed_loader.cc:
```diff
@@ -402,19 +402,31 @@ distributed_loader::make_sstables_available(sstables::sstable_directory& dir, sh
+struct is_shared {
+    using yes = std::true_type;
+    using no = std::false_type;
+};
```
How about using seastar::bool_class for this?

tchaikov replied:
Botond also had similar suggestion. see https://github.com/scylladb/scylladb/pull/13073#discussion_r1136591150. but for the reason noted in the referenced comment, an instance of `seastar::bool_class` cannot be used as a template parameter.

bhalevy replied:
I see.  Well using `sharded` we'll able to get rid of sharing and this won't be required.
That's another indication of the hidden complexity of using std::atomic for sharing across shards.

tchaikov replied:
if we agree on using a `sharded<generator>`, can we postpone the change to another PR? because, if the specified `shard_id` is not the one on which the local instance lives, we need to `invoke_on(shard_id)` to get the generation id from its peer. so the `sharded<generator>` change will not only change the generator's `operator()` to return a future, it will also change *all* the caller sites where a synchronized sstable creator is expected. that'd be more much than a refactor. to be more specific, currently, `compaction::_sstable_creator` is the destination of the the assignment chain of the `creator` parameter of `reshard(.., sstables::compaction_sstable_creator_fn creator, ...)` call. and all callers of `compaction::_sstable_creator` returns instance values. to change them and their callers, is non-trivial and risky.

bhalevy replied:
Other than the distributed loader and maybe sstable_directory, the other users are already sharded under replica::table so I don't see why they would need to create an sstable for another shard.

bhalevy replied:
that said, if we go the sharded way, I agree it deserves a separate PR, but as preparation for this one, not as a follow up.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/454/

tchaikov commented:
rebased against master HEAD.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/658/

denesb reviewed (changes requested):
This PR seems to do more than its title suggests as it also migrates sstables to UUID based generation. There is no mention about this in the cover-letter so its not clear if this is a complete migration or not. My suggestion is to defer this to a follow-up PR.

As for the generator part: I think we should make sure callers only call from local shard. Callers from remote shards (resharding) should use `smp::submit_to()` so we can avoid the atomic and the ugly `std::conditional_t` around it.
Also, let's move the generator into `generation_type` instead of moving it around withing the same PR.

tchaikov commented on replica/distributed_loader.cc:
```diff
@@ -436,6 +437,22 @@ sstables::shared_sstable make_sstable(replica::table& table, fs::path dir, sstab
+    //
+    // operator() is called by multiple shards in parallel when performing reshard,
+    // so we have to use atomic<> here.
+    std::atomic<sstables::generation_type::int_t> _last_generation;
```
as i explained in previous comment, this would change all the callers function signatures in distributed loader.

tchaikov replied:
> Why not put the generator in generation_type.hh in the first place?

the use case in distributed_loader.cc is more complicated. as it involves cross shard communication (it uses `atomic<>`) to address this problem, so i wanted to start with a refactor. but i am also happy to squash these changes, as long as that's preferred.

tchaikov replied:
> Aren't all calls to the sstable creator method made in a steastar::thread context? 

ahh, seems so, see https://github.com/scylladb/scylladb/blob/c501163f955e8f665211fb1456f92fcedf07dd3d/compaction/compaction.cc#L1664-L1680  . that's a relief!

> Would it be too dicey to rely on that? What do you think @raphaelsc?

tchaikov commented:
> This PR seems to do more than its title suggests as it also migrates sstables to UUID based generation. There is no mention about this in the cover-letter so its not clear if this is a complete migration or not. My suggestion is to defer this to a follow-up PR.
> 

sorry, i just pushed my local branch to this PR without removing the uuid changes.

> As for the generator part: I think we should make sure callers only call from local shard. Callers from remote shards (resharding) should use `smp::submit_to()` so we can avoid the atomic and the ugly `std::conditional_t` around it. Also, let's move the generator into `generation_type` instead of moving it around withing the same PR.

regarding the sharded change, my concern was your comments and Avi's  objection to the overkill design. but now that you are fine with it, i will see what i can do. please note, the cross-shard change would change the code all the way down to `compaction_sstable_creator_fn`.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/659/

denesb commented:
> regarding the sharded change, my concern was your comments and Avi's objection to the overkill design. but now that you are fine with it, i will see what i can do. please note, the cross-shard change would change the code all the way down to `compaction_sstable_creator_fn`.

I think the current state of each table object having a generator member is fine. No need to make it a separate sharded service.

tchaikov commented:
> > regarding the sharded change, my concern was your comments and Avi's objection to the overkill design. but now that you are fine with it, i will see what i can do. please note, the cross-shard change would change the code all the way down to `compaction_sstable_creator_fn`.
> 
> I think the current state of each table object having a generator member is fine. No need to make it a separate sharded service.

yes, each table/column family has its own generator. but i think the crux is the use case of distributed loader.

denesb commented:
> yes, each table/column family has its own generator. but i think the crux is the use case of distributed loader.

RIght. There, locally, we can use a `sharded<>` instead of `std::vector<std::unique_ptr>>` as we have right now. Or just have all of them on shard 0. Either works.

tchaikov commented:
changelog

- drop the uuid related changes. will add them in another PR
- instead of implement the generator via a refactory, just add the generator in a dedicated change
- use a `sharded<>` service in distributed_loader. as the reshard / reshape tasks are performed in a seastar::thread context, there is no need to futurize the functions on the call stack which creates a new generation identifier. this change is trivial enough to include this PR.

@denesb hi Botond, could you please take another look?

denesb commented on replica/distributed_loader.cc:
```diff
@@ -465,22 +462,34 @@ distributed_loader::process_upload_dir(distributed<replica::database>& db, distr
+                    return gg.next(shard);
+                });
+            }
+        };
```
This could be much simpler.
Just have `sharded<sstables::sstable_generation_generator> sharded_generator` and in `make_sstable()` below, you call `auto generation = sharded_generator.invoke_on(shard, [] { ... ]).get();`.

tchaikov replied:
thanks! incorporated the proposed change.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/675/

tchaikov commented:
changelog

- instead of using `peering_sharded_service`, use plain `sharded<>` by incorporating @denesb 's suggestion
- comment on `sharded_parameter` for posterity
- comment on `sstable_generation_generator::operator()` to explain why we are incrementing the number by `smp::count`.

tchaikov commented:
https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/710/#showFailuresLink quite a few tests failed. looking.

tchaikov commented:
changelog

- fixed `table::update_sstables_known_generation()`. should have use different `normalized_gen` for different shard.

@denesb hi Botond, could you take another look?

raphaelsc commented on replica/distributed_loader.cc:
```diff
@@ -465,22 +462,24 @@ distributed_loader::process_upload_dir(distributed<replica::database>& db, distr
+                // get the base generation so we can increment it by smp::count without
+                // conflicting with other shards
+                return highest_generation.value() / smp::count * smp::count + this_shard_id();
+            })).get();
```
I think this logic should live in sstables::sstable_generation_generator. It's irrelevant to UUID.

Also note that value / smp::count * smp::count + this_shard_id() will align down.

we need to add smp::count to value, before aligning it down.

tchaikov replied:
please see the implementation of the generator, we always add smp::count before returning the next id. so no need to add smp::count.

raphaelsc replied:
Alright. But please let's move all this arithmetic there. It's generation type specific. An implementation detail that the user doesn't have to be aware of.

raphaelsc commented on replica/table.cc:
```diff
@@ -66,30 +66,25 @@ static seastar::metrics::label keyspace_label("ks");
-        _sstable_generation.emplace(normalized_generation);
-        tlogger.debug("{}.{} updated highest known generation to {}", schema()->ks_name(), schema()->cf_name(), *_sstable_generation);
+    auto normalized_gen = gen - gen % smp::count + this_shard_id();
+    if (_sstable_generation_generator) {
```
With "normalization" moved to sstables::sstable_generation_generator, we can get rid of the arithmetic above too. User of the generator doesn't have to know about this. That's implementation detail pushed upwards.

raphaelsc reviewed (changes requested):
Looks good overall, we just need to hide implementation detail from sstables::sstable_generation_generator user. Left two comments about it.

tchaikov commented:
changelog

- push the normalization logic down to generation generator

@raphaelsc hi Raph, thanks for your suggestion. updated accordingly.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/802/

raphaelsc reviewed (approved):
looks good. thanks!

denesb commented:
Queued.
